README.md
# nagios-plugins-vshn-icinga2

Currently the checks are written in Ruby and are proof-of-concept quality at best. Input validation and exception handling is very basic or not existing. Use at own risk!

The idea is to rewrite the scripts in Python or Go and using a Icinga2 API library.

## `icinga2internals`

Checks some Icinga2 internal metrics for IDO, InfluxdbWriter, number of API endpoints, etc. As the check just uses the API, it can also check remote icinga2 instances.

High queues for IDO or InfluxDB usually indicate that the InfluxDB or the MariaDB server is not able to catchup with the load or the connection is broken otherwise. High queues can lead to icinga2 behaving strangely over time and data might get lost.
```
Usage: ./icinga2internals.rb [options]
    -H, --host=host
    -P, --port=port
    -u, --user=username
    -p, --password=password
    -w, --idowarn=value # default: 500
    -c, --idocrit=value # default: 1000
    -g, --influxwarn=value # default: 500
    -s, --influxcrit=value # default: 1000
```
### Example
```
./icinga2internals.rb -H <icingahost> -u <user> -p <key>
```
Output examples:
```
OK: ido_query_queue_items OK: 0, ido_query_queue_item_rate OK: 630, influx_data_buffer_items OK: 0, influx_work_queue_item_rate OK: 318, influx_work_queue_items OK: 19, num_not_conn_endpoints OK: 0
```
```
CRITICAL: influx_data_buffer_items NOT OK: 1014 greater 1000
```

## `submit_remote`

Can be use with any Icinga2 / Nagios check script to submit the result as a passive check result to any Icinga2 instance via the API. A API user with sufficient permissions is needed.

It basically runs the command specified by `-c` and submits the results so the host and service specified by `-s`.

```
Usage: ./submit_remote.rb [options]
    -H, --host=MANDATORY
    -P, --port=MANDATORY
    -u, --user=MANDATORY
    -p, --password=MANDATORY
    -c, --command=MANDATORY
    -s, --remoteservice=MANDATORY # example: hostname!servicename
```

### Examples
This one checks the icinga2 internals (see above) and forwards the results to a remote icinga2. The check also passes through the results to the local Icinga2, so it's basically transparent.
```
./submit_remote.rb -H <remote icinga host> -s 'icinga-server.vagrant.dev!icinga2_test.remote.vagrant.dev' -u <remote user> -p <remote pass> -c "./icinga2internals.rb -H localhost -u <local user> -p <local pass>"
```

## Example Icinga2 config
service on the remote icinga2 with "heartbeat" / freshness check:
```
object Service "icinga2_test.remote.vagrant.dev" {
  check_command = "dummy"
  check_interval = 10s

  /* Set the state to UNKNOWN (3) if freshness checks fail. */
  vars.dummy_state = 2

  /* Use a runtime function to retrieve the last check time and more details. */
  vars.dummy_text = {{
    var service = get_service(macro("$host.name$"), macro("$service.name$"))
    var lastCheck = DateTime(service.last_check).to_string()

    return "No check results received. Last result time: " + lastCheck
  }}

  host_name = "icinga-server.vagrant.dev"
}
```
local check:
```
object Service "icinga2internals" {
  check_command = "submit_remote"
  check_interval = 20s
  display_name = "Icinga2 internal metrics"

  vars.submit_remote_host = "localhost"
  vars.submit_remote_user = "user"
  vars.submit_remote_password = "pass"
  vars.submit_remote_remoteservice = "icinga-server.vagrant.dev!icinga2_test.remote.vagrant.dev"
  vars.submit_remote_command = PluginDir + "/icinga2internals.rb -H localhost -u <local user> -p <local pass>

  host_name = "icinga-server.vagrant.dev"
}
```


debian/changelog
nagios-plugins-vshn-icinga2 (0.3.0) xenial; urgency=medium

  * Add submit_remote and icinga2inernals checks 

 -- Marco Fretz <marco.fretz@vshn.ch>  Wed, 14 Mar 2018 13:05:48 +0000

nagios-plugins-vshn-icinga2 (0.2.1) xenial; urgency=medium

  * Filter services from down hosts 

 -- Marco Fretz <marco.fretz@vshn.ch>  Wed, 28 Feb 2018 10:02:35 +0000

nagios-plugins-vshn-icinga2 (0.2.0) xenial; urgency=medium

  * Only use hard states 

 -- Marco Fretz <marco.fretz@vshn.ch>  Tue, 23 Jan 2018 10:16:30 +0000

nagios-plugins-vshn-icinga2 (0.1.2) xenial; urgency=medium

  * fix section 

 -- Marco Fretz <marco.fretz@vshn.ch>  Fri, 19 Jan 2018 10:06:22 +0000

nagios-plugins-vshn-icinga2 (0.1.1) xenial; urgency=medium

  * fix 

 -- Marco Fretz <marco.fretz@vshn.ch>  Fri, 19 Jan 2018 09:58:15 +0000

nagios-plugins-vshn-icinga2 (0.1.0) xenial; urgency=medium

  * Initial release.

 -- Marco Fretz <marco.fretz@vshn.ch>  Fri, 19 Jan 2018 09:40:30 +0000


icinga2stats.rb
#!/usr/bin/ruby

require 'net/http'
require 'uri'
require 'json'
require 'optparse'

options = {}
options[:port] = 5665
options[:managed_var] = 'vshnmanaged'

OptionParser.new do |opts|
  opts.banner = "Usage: icinga2stats.rb [options]"
  opts.on("-H", "--host=host" , String) do |v|
    options[:host] = v;
  end
  opts.on("-P", "--port=port", Integer) do |v|
    options[:port] = v;
  end
  opts.on("-u", "--user=username", String) do |v|
    options[:user] = v;
  end
  opts.on("-p", "--password=password", String) do |v|
    options[:pass] = v;
  end
  opts.on("-f", "--filter=filter", String) do |v|
    options[:managed_var] = v;
  end
end.parse!

datapoints = {
  'num_hosts_up' => {
    'filter' => 'host.state==0 && host.vars.' + options[:managed_var].to_s + '!=false',
    'url' => '/v1/objects/hosts',
  },
  'num_hosts_down' => {
    'filter' => 'host.state_type==1 && host.state!=0 && host.downtime_depth==0 && host.acknowledgement==0' +
      '&& host.vars.' + options[:managed_var].to_s + '!=false',
    'url' => '/v1/objects/hosts',
  },
  'num_services_ok_filtered' => {
    'filter' => 'service.state==0 && service.vars.' + options[:managed_var].to_s + '==false',
    'url' => '/v1/objects/services',
  },
  'num_services_critical' => {
    'filter' => 'service.state_type==1 && service.state==2 && service.downtime_depth==0 && service.acknowledgement==0 && host.state==0' +
      '&& service.vars.' + options[:managed_var].to_s + '!=false',
    'url' => '/v1/objects/services',
  },
  'num_services_warning' => {
    'filter' => 'service.state_type==1 && service.state==1 && service.downtime_depth==0 && service.acknowledgement==0 && host.state==0' +
      '&& service.vars.' + options[:managed_var].to_s + '!=false',
    'url' => '/v1/objects/services',
  }
}

header = {
  "Accept" => "application/json",
  "X-HTTP-Method-Override" => "GET"
}
begin
  # Create the HTTP objects
  http = Net::HTTP.start(options[:host], options[:port], :use_ssl => true, :verify_mode => 0)
  results = {}

  datapoints.each do |key, value|
    data = {
      'filter' => value['filter'],
      'attrs' => [ 'name' ]
    }
    request = Net::HTTP::Post.new(value['url'], header)
    request.basic_auth options[:user], options[:pass]
    request.body = data.to_json
    response = http.request(request)
    output = JSON.parse(response.body)
    results[key] = output['results'].length
  end

  # get okay services the quick way (all ok minus filtered)
  request = Net::HTTP::Get.new('/v1/status/CIB', header)
  request.basic_auth options[:user], options[:pass]
  response = http.request(request)
  output = JSON.parse(response.body)
  services_ok = output['results'][0]['status']['num_services_ok']

  results['num_services_ok'] = services_ok - results['num_services_ok_filtered']

  perf_data = []
  results.each do |key, value|
    perf_data.push(key + "=" + value.to_s)
  end

  puts 'OK: Stats collected |' + perf_data.join(' ')
  exit 0
rescue
  puts 'UNKNOWN: An error occured, please run the check manually to see what it wrong'
  exit 3
end


submit_remote.rb
#!/usr/bin/ruby

require 'net/http'
require 'uri'
require 'json'
require 'optparse'

options = {}
options[:port] = 5665

OptionParser.new do |opts|
  opts.banner = "Usage: #{$0} [options]"
  opts.on("-H", "--host=host", "=MANDATORY", String) do |v|
    options[:host] = v;
  end
  opts.on("-P", "--port=port", "=MANDATORY", Integer) do |v|
    options[:port] = v;
  end
  opts.on("-u", "--user=username", "=MANDATORY", String) do |v|
    options[:user] = v;
  end
  opts.on("-p", "--password=password", "=MANDATORY", String) do |v|
    options[:pass] = v;
  end
  opts.on("-c", "--command=command", "=MANDATORY", String) do |v|
    options[:command] = v;
  end
  opts.on("-s", "--remoteservice=host!service", "=MANDATORY", String) do |v|
    options[:remoteservice] = v;
  end
end.parse!
raise ArgumentError.new("Wrong number of options") unless options.keys.count == 6

begin
  cmd = options[:command]
  cmd_result = %x(#{cmd})
  cmd_code = $?.exitstatus

  unless cmd_result.empty?
    http = Net::HTTP.start(options[:host], options[:port], :use_ssl => true, :verify_mode => 0)
    data = {
      'exit_status' => cmd_code,
      'plugin_output' => cmd_result,
    }
    url = '/v1/actions/process-check-result?service=' + options[:remoteservice]
    request = Net::HTTP::Post.new(url, { 'Accept' => 'application/json' })
    request.basic_auth options[:user], options[:pass]
    request.body = data.to_json
    response = http.request(request)
    if response.code != '200'
      puts 'UNKNOWN: API call to ' + options[:host] + ' failed: ' + response.code + ' ' + response.message
      exit 3
    end
    output = JSON.parse(response.body)
    puts cmd_result
    exit cmd_code
  else
    puts 'UNKNOWN: No output from command. Not submitting to remote icinga.'
    exit 3
  end
rescue
  puts "UNKNOWN: Exception in #{$0}"
  exit 3
end


icinga2internals.rb
#!/usr/bin/ruby

require 'net/http'
require 'uri'
require 'json'
require 'optparse'

options = {}
results = {}
states = { 0 => 'OK', 1 => 'WARNING', 2 => 'CRITICAL', 3 => 'UNKNOWN' }
options[:port] = 5665
options[:ido_warn] = 500
options[:ido_crit] = 1000
options[:influx_warn] = 500;
options[:influx_crit] = 1000;

OptionParser.new do |opts|
  opts.banner = "Usage: #{$0} [options]"
  opts.on("-H", "--host=host" , String) do |host|
    options[:host] = host;
  end
  opts.on("-P", "--port=port", Integer) do |v|
    options[:port] = v;
  end
  opts.on("-u", "--user=username", String) do |v|
    options[:user] = v;
  end
  opts.on("-p", "--password=password", String) do |v|
    options[:pass] = v;
  end
  opts.on("-w", "--idowarn=value", Integer) do |v|
    options[:ido_warn] = v;
  end
  opts.on("-c", "--idocrit=value", Integer) do |v|
    options[:ido_crit] = v;
  end
  opts.on("-g", "--influxwarn=value", Integer) do |v|
    options[:influx_warn] = v;
  end
  opts.on("-s", "--influxcrit=value", Integer) do |v|
    options[:influx_crit] = v;
  end
end.parse!
raise ArgumentError.new("Wrong number of options") unless options.keys.count == 8

def get_status_values(status_object, name, key)
  # only get the first instance, multiple instnaces of ido / influx, etc. not supported atm
  if a = status_object['results'].select {|key| key['name'] == name}[0]
    return a['status'][key].first[1]
  end
end

def compare_metric(current, warn, crit, name)
  if current.round > crit
    return { :code => 2, :text => name + ' NOT OK: ' + current.round.to_s + ' > ' + crit.to_s }
  elsif current.round > warn
    return { :code => 1, :text => name + ' NOT OK: ' + current.round.to_s + ' > ' + warn.to_s }
  end
  return { :code => 0, :text => name + ' OK: ' + current.round.to_s }
end

header = {
  "Accept" => "application/json",
  "X-HTTP-Method-Override" => "GET"
}
begin
  # Create the HTTP objects
  http = Net::HTTP.start(options[:host], options[:port], :use_ssl => true, :verify_mode => 0)
  request = Net::HTTP::Get.new('/v1/status', header)
  request.basic_auth options[:user], options[:pass]
  response = http.request(request)
  if response.code != '200'
    puts 'UNKNOWN: API call to ' + options[:host] + ' failed: ' + response.code + ' ' + response.message
    exit 3
  end
  status_object = JSON.parse(response.body)

  if influx_status = get_status_values(status_object, 'InfluxdbWriter', 'influxdbwriter')
    results['influx_data_buffer_items'] = compare_metric(influx_status['data_buffer_items'], options[:influx_warn], options[:influx_crit], 'influx_data_buffer_items')
    results['influx_work_queue_item_rate'] = compare_metric(influx_status['work_queue_item_rate'], options[:influx_warn], options[:influx_crit], 'influx_work_queue_item_rate')
    results['influx_work_queue_items'] = compare_metric(influx_status['work_queue_items'], options[:influx_warn], options[:influx_crit], 'influx_work_queue_items')
  end
  ido_status = get_status_values(status_object, 'IdoMysqlConnection', 'idomysqlconnection')
  app_status = get_status_values(status_object, 'IcingaApplication', 'icingaapplication')
  # status for ApiListener uses a different format / does not has multiple instances
  api_status = status_object['results'].select {|key| key['name'] == 'ApiListener'}[0]['status']['api']

  results['ido_query_queue_items'] = compare_metric(ido_status['query_queue_items'], options[:ido_warn], options[:ido_crit], 'ido_query_queue_items')
  results['ido_query_queue_item_rate'] = compare_metric(ido_status['query_queue_item_rate'], options[:ido_warn], options[:ido_crit], 'ido_query_queue_item_rate')
  results['api_num_not_conn_endpoints'] = compare_metric(
    api_status['num_not_conn_endpoints'],
    (api_status['num_conn_endpoints'] * 0.2).round,
    (api_status['num_conn_endpoints'] * 0.5).round,
    'num_not_conn_endpoints'
  )

  final_code = results.values.map { |i| i[:code] }.max
  if final_code == 0
    puts 'OK: ' + results.map{|k, v| v[:text]}.join(', ')
  else
    puts states[final_code] + ': ' + results.select {|k, v| v[:code] > 0}.map{|k, v| v[:text]}.join(', ')
  end
rescue
   puts "UNKNOWN: Exception in #{$0}"
   exit 3
end
exit final_code
